stratum/hal/lib/phal/tai/tai_wrapper/tai_manager.h

```diff
@@ -0,0 +1,280 @@
+  static bool SetValueToTAIAttribute(TAIAttribute* tai_attribute,
+                                     const T& value_to_set);
+  template <typename T>
+  static T TaiAttributeToResponse(const TAIAttribute& attribute);
+
+ protected:
+  explicit TAIManager(std::unique_ptr<TAIWrapperInterface> wrapper);
+
+ private:
+  // The "TAIWrapper" class manages TAI objects' lifetime, as well as the access
+  // to them.
+  // Now, to experience a thread-safe TAI manager interaction, this mutex should
+  // be LOCKED BEFORE
+  //
+  //   * any TAIManager method call
+  //
+  // and UNLOCKED AFTER
+  //
+  //   * no TAI manager methods calls follow next in the scope of some single
+  //     action (e.g., get/set a single attribute);
+  //   * all TAIObject pointers retrieved from the TAIManager are released.
+  //
+  //  That means, if we retrieve a weak_ptr from the TAIManager, we MUST NOT
+  //  unlock the mutex until the pointer is released.
+  //
+  mutable absl::Mutex tai_wrapper_mutex_;
+  std::unique_ptr<TAIWrapperInterface> tai_wrapper_;
+
+  // RW mutex lock for protecting the singleton instance initialization and
+  // reading it back from other threads. Unlike other singleton classes, we
+  // use RW lock as we need the pointer to class to be returned.
+  static absl::Mutex init_lock_;
+
+  // The singleton instance.
+  static TAIManager* singleton_;
+  GUARDED_BY(init_lock_);
+};
+
+/*!
+ * \brief TAIManager::GetValue method make TAI get value by \param attr_id TAI
+ * attribute
+ * \param attr_id it's TAI attribute from what value will be getted
+ * \param module_netif_pair it's pair of module id and related to it networkif
+ * id interface
+ * \return valid value or invalid ::util::Status
+ * \note method is thread safe a thread-safe.
+ */
+template <typename T>
+::util::StatusOr<T> TAIManager::GetValue(
+    const tai_attr_id_t& attr_id,
+    const std::pair<uint64, uint32>& module_netif_pair) const {
+  absl::ReaderMutexLock wrapper_lock(&tai_wrapper_mutex_);
+
+  const std::shared_ptr<TAIObject> tai_object =
+      tai_wrapper_->GetObject(TAIPathValidator::NetworkPath(module_netif_pair))
+          .lock();
+  if (!tai_object) {
+    std::stringstream error_msg;
+    error_msg << "Location of module id: " << module_netif_pair.first
+              << " or network interface id: " << module_netif_pair.second
+              << " is not valid";
+
+    LOG(ERROR) << error_msg.str();
+    RETURN_ERROR(ERR_INTERNAL) << error_msg.str();
+  }
+
+  // Retrieve the requested attribute from the TAI object.
+  tai_status_t return_code;
+  const TAIAttribute tai_attr = tai_object->GetAttribute(attr_id, &return_code);
+  if (return_code != TAI_STATUS_SUCCESS) {
+    std::stringstream error_msg;
+    error_msg << "Can't get the attribute. TAI return code: " << return_code;
+
+    LOG(ERROR) << error_msg.str();
+    RETURN_ERROR(ERR_INTERNAL) << error_msg.str();
+  }
+
+  return ::util::StatusOr<T>(TaiAttributeToResponse<T>(tai_attr));
+}
```

Comment:
move implememtations to `.cc` file, same as below

Reply:
This is the template method. We should either define it in the header file or directly include the .cpp file with its implementation wherever we use it. What approach should we follow in this case?

Reply:
I think we missed the template. It's ok to leave as is for now.